mlflow/gateway/providers/bedrock.py

```diff
@@ -0,0 +1,279 @@
+import json
+import os
+import time
+from enum import Enum
+
+import boto3
+import botocore.config
+import botocore.exceptions
+from fastapi import HTTPException
+from fastapi.encoders import jsonable_encoder
+
+from mlflow.gateway.config import AWSBedrockConfig, AWSIdAndKey, AWSRole, RouteConfig
+from mlflow.gateway.constants import (
+    MLFLOW_AI_GATEWAY_ANTHROPIC_DEFAULT_MAX_TOKENS,
+    MLFLOW_AI_GATEWAY_ANTHROPIC_MAXIMUM_MAX_TOKENS,
+)
+from mlflow.gateway.exceptions import AIGatewayConfigException
+from mlflow.gateway.providers.base import BaseProvider
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+
+
+class AWSBedrockModelProvider(Enum):
+    AMAZON = "amazon"
+    COHERE = "cohere"
+    AI21 = "ai21"
+    ANTHROPIC = "anthropic"
+
+    @classmethod
+    def of_str(cls, name: str):
+        name = name.lower()
+
+        for opt in cls:
+            if opt.name.lower() == name or opt.value.lower() == name:
+                return opt
+
+
+class AWSBedrockProvider(BaseProvider):
+    def __init__(self, config: RouteConfig):
+        super().__init__(config)
+
+        if config.model.config is None or not isinstance(config.model.config, AWSBedrockConfig):
+            raise TypeError(f"Invalid config type {config.model.config}")
+        self.bedrock_config: AWSBedrockConfig = config.model.config
+        self._client = None
+        self._client_created = 0
+
+    def _client_expired(self):
+        if not isinstance(self.bedrock_config.aws_config, AWSRole):
+            return False
+
+        return (
+            (time.monotonic_ns() - self._client_created)
+            >= (self.bedrock_config.aws_config.session_length_seconds) * 1_000_000_000,
+        )
+
+    def get_bedrock_client(self):
```

Comment:
This function is a bit hard to read and contains a lot of branching logic that may be hard to troubleshoot and test. Could this be refactored in a manner like this?

```python

def get_bedrock_client(self):
    if self._client is not None and not self._client_expired():
        return self._client

    session_args = self._construct_session_args()
    session = boto3.Session(**session_args)

    client_args = self._construct_client_args(session, self.bedrock_config)

    try:
        self._client, self._client_created = (
            session.client(service_name="bedrock", **client_args),
            time.monotonic_ns(),
        )
        return self._client
    except botocore.exceptions.UnknownServiceError as e:
        raise AIGatewayConfigException(
            "Cannot create AWS Bedrock client; ensure boto3/botocore "
            "linked from the AWS Bedrock user guide are installed. "
            "Otherwise likely missing credentials or accessing account without to "
            "AWS Bedrock Private Preview"
        ) from e

# Allow for isolating the session arguments
def _construct_session_args(self):
    aws_config = self.bedrock_config.aws_config
    region_name = aws_config.aws_region or os.getenv("AWS_REGION", os.getenv("AWS_DEFAULT_REGION"))
    profile_name = os.getenv("AWS_PROFILE")

    return {k: v for k, v in {
        "region_name": region_name,
        "profile_name": profile_name
    }.items() if v}

# simplify the conditional logic and use dict.update() instead of direct assignment
def _construct_client_args(self, session, aws_config):
    client_args = {}

    if isinstance(aws_config, AWSRole):
        role = session.client(service_name="sts").assume_role(
            RoleArn=aws_config.aws_role_arn,
            RoleSessionName="ai-gateway-bedrock",
            DurationSeconds=aws_config.session_length_seconds,
        )
        client_args.update({
            "aws_access_key_id": role["Credentials"]["AccessKeyId"],
            "aws_secret_access_key": role["Credentials"]["SecretAccessKey"],
            "aws_session_token": role["Credentials"]["SessionToken"]
        })
    elif isinstance(aws_config, AWSIdAndKey):
        client_args.update({
            "aws_access_key_id": aws_config.aws_access_key_id,
            "aws_secret_access_key": aws_config.aws_secret_access_key,
            "aws_session_token": aws_config.aws_session_token
        })
    # remove the else block (unreachable code for the pass, which does nothing)

    return client_args
```

Reply:
Sure, that looks good.  Not sure that dict.update adds a lot there, but it will look nice when you guys drop 3.7 and convert all those to `|=` :p